src/aliceVision/image/caching.hpp

```diff
@@ -0,0 +1,435 @@
+// This file is part of the AliceVision project.
+// Copyright (c) 2022 AliceVision contributors.
+// This Source Code Form is subject to the terms of the Mozilla Public License,
+// v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Image.hpp"
+#include "pixelTypes.hpp"
+#include "io.hpp"
+#include "imageAlgo.hpp"
+
+#include <aliceVision/system/Logger.hpp>
+
+#include <boost/filesystem.hpp>
+#include <boost/functional/hash.hpp>
+
+#include <memory>
+#include <unordered_map>
+#include <functional>
+#include <list>
+#include <mutex>
+#include <thread>
+
+
+namespace aliceVision {
+namespace image {
+
+/**
+ * @brief A struct used to identify a cached image using its file description, color type info and half-sampling level.
+ */
+struct CacheKey 
+{
+    std::string filename;
+    int nbChannels;
+    oiio::TypeDesc::BASETYPE typeDesc;
+    int halfSampleLevel;
```

Comment:
downsamplingLevel

Reply:
I think `halfSampleLevel` is more explicit than `downscaleLevel` because the actual downscale that we apply to the cached images is $2^{halfSampleLevel}$ (and we actually use the word `downscale` in several places to refer to this power of 2).

Reply:
Is there a reason why we only want to allow downscaling in the image cache with power of two level?
Managing all downscale levels will be useful (e.g. depth maps estimation).

Reply:
Now that you say it, I don't see why not 😄 
I'll put the PR back in draft in the meantime, that will require some testing

Reply:
Done !